substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/JavaUtilZipSubstitutions.java

```diff
@@ -209,62 +228,111 @@ private int deflateBytes(long addr, byte[] b, int off, int len, int flush) {
         }
     }
 
+    // (JDK 11)src/java.base/share/native/libzip/Deflater.c-static jlong doDeflate(JNIEnv *env, jobject this, jlong addr,
+    //                       jbyte *input, jint inputLen,
+    //                       jbyte *output, jint outputLen,
+    //                       jint flush, jint params)
+    private static long doDeflate(long addr, CCharPointer input, int inputLen, CCharPointer output, int outputLen, int flush, int params) {
+        z_stream strm = WordFactory.pointer(addr);
+        int inputUsed, outputUsed;
+        int finished = 0;
+        int setParams = params & 1;
+
+        strm.set_next_in(input);
+        strm.set_next_out(output);
+        strm.set_avail_in(inputLen);
+        strm.set_avail_out(outputLen);
+
+        if (setParams != 0) {
+            int strategy = (params >> 1) & 3;
+            int level = params >> 3;
+            int res = ZLib.deflateParams(strm, level, strategy);
+            if (res == ZLib.Z_OK() || res == ZLib.Z_BUF_ERROR()) {
+                if (res == ZLib.Z_OK()) {
+                    setParams = 0;
+                }
+            } else {
+                throw new InternalError();
```

Comment:
What happened to `strm->msg`?  If the message is not useful, then add a comment saying you are deliberately ignoring it.